apt went "missing"

**Describe the bug**
Between bdce0b36b28f3750a8993316850ccaadf428a068 and 6c45ecfdbee86ed80f858717901b20b90ec31dbc, apt somehow got reintroduced into the repository ([pkgs/tools/package-management/apt](https://github.com/NixOS/nixpkgs/commit/d6aeae8f90a4935525915431e3b08ebf1b7d5bf3)) but without an entry in all-packages.nix.

- It could be bumped to 2.1.18 - Out-of-date package report
- It doesn't compile for the same reason for which it was originally removed from Nixpkgs and I don't know how to fix it - bug report
  ```
  [ 33%] Building CXX object apt-pkg/CMakeFiles/apt-pkg.dir/acquire-item.cc.o
  In file included from /nix/store/kg2s3k77ifsyd59ac08mjmmjjmsvgsvg-gcc-10.2.0/include/c++/10.2.0/ext/string_conversions.h:41,
                   from /nix/store/kg2s3k77ifsyd59ac08mjmmjjmsvgsvg-gcc-10.2.0/include/c++/10.2.0/bits/basic_string.h:6535,
                   from /nix/store/kg2s3k77ifsyd59ac08mjmmjjmsvgsvg-gcc-10.2.0/include/c++/10.2.0/string:55,
                   from /build/source/build/include/apt-pkg/hashes.h:18,
                   from /build/source/build/include/apt-pkg/acquire.h:68,
                   from /build/source/build/include/apt-pkg/acquire-item.h:22,
                   from /build/source/apt-pkg/acquire-item.cc:17:
  /nix/store/kg2s3k77ifsyd59ac08mjmmjjmsvgsvg-gcc-10.2.0/include/c++/10.2.0/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
     75 | #include_next <stdlib.h>
        |               ^~~~~~~~~~
  compilation terminated.
  make[2]: *** [apt-pkg/CMakeFiles/apt-pkg.dir/build.make:82: apt-pkg/CMakeFiles/apt-pkg.dir/acquire-item.cc.o] Error 1
  ```
- I need apt for another package to work properly (apt -> python-apt -> python-debian -> click: https://github.com/NixOS/nixpkgs/pull/108366) - packaging request?

**To Reproduce**
1. [Search](https://search.nixos.org/packages?channel=unstable&query=apt) for apt package -> not packaged
2. [Search](https://github.com/NixOS/nixpkgs/search?q=apt&type=code) in Nixpkgs repo -> but package code exists?
3. Check commit/PR history -> #68029 (staging) removed all apt code, #69571 (master) bumped apt

**Expected behavior**
apt being either in all-packages.nix or missing from the repository completely. Preferably, a working apt derivation.

~~**Screenshots**~~

**Additional context**
https://github.com/NixOS/nixpkgs/pull/108366

**Notify maintainers**
@cstrahan, package maintainer for apt
@georgyo, last bump to package
@fpletz, originally removed the package

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.9.16, NixOS, 20.09.2642.88f00e7e12d (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-20.09.2642.88f00e7e12d"`
 - channels(bt1cn): `"unstable-21.03pre259928.2f47650c2f2"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: [ apt ]
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/tools/misc/triehash/default.nix
+++ b/pkgs/tools/misc/triehash/default.nix
@@ -1,0 +1,48 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, perlPackages
+}:
+
+stdenv.mkDerivation rec {
+  pname = "triehash";
+  version = "0.3";
+
+  src = fetchFromGitHub {
+    owner = "julian-klode";
+    repo = pname;
+    rev = "debian/0.3-3";
+    hash = "sha256-LxVcYj2WKHbhNu5x/DFkxQPOYrVkNvwiE/qcODq52Lc=";
+  };
+
+  nativeBuildInputs = [
+    perlPackages.perl
+  ];
+
+  postPatch = ''
+    patchShebangs triehash.pl
+  '';
+
+  dontConfigure = true;
+
+  dontBuild = true;
+
+  installPhase = ''
+    runHook preInstall
+
+    install -d $out/bin $out/share/doc/${pname}/ $out/share/${pname}/
+    install triehash.pl $out/bin/triehash
+    install README.md $out/share/doc/${pname}/
+    cp -r tests/ $out/share/${pname}/tests/
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    homepage = "https://github.com/julian-klode/triehash";
+    description = "Order-preserving minimal perfect hash function generator";
+    license = with licenses; mit;
+    maintainers = with maintainers; [ AndersonTorres ];
+    platforms = perlPackages.perl.meta.platforms;
+  };
+}

--- a/pkgs/tools/package-management/apt/default.nix
+++ b/pkgs/tools/package-management/apt/default.nix
@@ -1,50 +1,84 @@
-{ stdenv, lib, fetchurl, pkg-config, cmake, perlPackages, curl, gtest
-, gnutls, libtasn1, xz, bzip2, lz4, zstd, libseccomp, udev
-, db, dpkg, libxslt, docbook_xsl, docbook_xml_dtd_45
-
-# used when WITH_DOC=ON
-, w3m
-, doxygen
-
-# used when WITH_NLS=ON
-, gettext
-
-# opts
-, withDocs ? true
-, withNLS ? true
+{ lib
+, stdenv
+, fetchurl
+, bzip2
+, cmake
+, curl
+, db
+, docbook_xml_dtd_45
+, docbook_xsl
+, dpkg
+, gnutls
+, gtest
+, libgcrypt
+, libseccomp
+, libtasn1
+, libxslt
+, lz4
+, perlPackages
+, pkg-config
+, triehash
+, udev
+, xxHash
+, xz
+, zstd
+, withDocs ? true , w3m, doxygen
+, withNLS ? true , gettext
 }:
 
 stdenv.mkDerivation rec {
   pname = "apt";
-  version = "1.8.4";
+  version = "2.3.8";
 
   src = fetchurl {
     url = "mirror://debian/pool/main/a/apt/apt_${version}.tar.xz";
-    sha256 = "0gn4srqaaym85gc8nldqkv01477kdwr136an2nlpbdrsbx3y83zl";
+    hash = "sha256-SFrxQwx14xWLcV5EJNv5bRtWQdxNzMUPVxssd5qDfyw=";
   };
 
-  nativeBuildInputs = [ pkg-config cmake gtest libxslt.bin ];
+  nativeBuildInputs = [
+    cmake
+    gtest
+    libxslt.bin
+    pkg-config
+    triehash
+  ];
 
   buildInputs = [
-    perlPackages.perl curl gnutls libtasn1 xz bzip2 lz4 zstd libseccomp udev db dpkg
+    bzip2
+    curl
+    db
+    dpkg
+    gnutls
+    libgcrypt
+    libseccomp
+    libtasn1
+    lz4
+    perlPackages.perl
+    udev
+    xxHash
+    xz
+    zstd
   ] ++ lib.optionals withDocs [
-    doxygen perlPackages.Po4a w3m docbook_xml_dtd_45
+    docbook_xml_dtd_45
+    doxygen
+    perlPackages.Po4a
+    w3m
   ] ++ lib.optionals withNLS [
     gettext
   ];
 
   cmakeFlags = [
-    "-DBERKELEY_DB_INCLUDE_DIRS=${db.dev}/include"
-    "-DGNUTLS_INCLUDE_DIR=${gnutls.dev}/include"
+    "-DBERKELEY_INCLUDE_DIRS=${db.dev}/include"
     "-DDOCBOOK_XSL=${docbook_xsl}/share/xml/docbook-xsl"
+    "-DGNUTLS_INCLUDE_DIR=${gnutls.dev}/include"
     "-DROOT_GROUP=root"
-    "-DWITH_DOC=${if withDocs then "ON" else "OFF"}"
     "-DUSE_NLS=${if withNLS then "ON" else "OFF"}"
+    "-DWITH_DOC=${if withDocs then "ON" else "OFF"}"
   ];
 
   meta = with lib; {
-    description = "Command-line package management tools used on Debian-based systems";
     homepage = "https://salsa.debian.org/apt-team/apt";
+    description = "Command-line package management tools used on Debian-based systems";
     license = licenses.gpl2Plus;
     platforms = platforms.linux;
     maintainers = with maintainers; [ cstrahan ];

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -31433,6 +31433,8 @@ with pkgs;
 
   emu2 = callPackage ../misc/emulators/emu2 { };
 
+  apt = callPackage ../tools/package-management/apt { };
+
   dpkg = callPackage ../tools/package-management/dpkg { };
 
   dumb = callPackage ../misc/dumb { };

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -19018,6 +19018,8 @@ with pkgs;
 
   tremor = callPackage ../development/libraries/tremor { };
 
+  triehash = callPackage ../tools/misc/triehash { };
+
   trillian = callPackage ../tools/misc/trillian {
     buildGoModule = buildGo115Module;
   };

